### ISSUE TITLE
Enhancement Request: pgformatter style formatting  

### ISSUE THREAD

**Issue by lav-patel** [dbeaver/dbeaver#17600]
**Is your feature request related to a problem? Please describe.**
--> We use [pgformatter ](https://www.postgresql.org/about/news/pgformatter-v50-released-2166/) style formatting for SQL as It is suggested by official pg docs, but it is frustrating to use another tool and reformat the SQL. 

**Describe the solution you'd like**
-->Ideally, I would like to have an additional option when I right-click and go to Format -> "Formate SQL (pgformatter)"

**Describe alternatives you've considered**
--> Currently, it is a two-step process, write SQL in dbeaver and make sure it works, put it in VS Code to format the code and bring the code back to d beaver.

**Additional context**
--> FYI: [pgformatter ](https://www.postgresql.org/about/news/pgformatter-v50-released-2166/) uses perl, so it may be hard to integrate in all os. (Ubuntu comes with perl)

**Comment by ShadelessFox** [comment]
Hello @lav-patel,

You can define an external formatter in <kbd>Preferences</kbd> &rArr; <kbd>Editors</kbd> &rArr; <kbd>SQL Editor</kbd> &rArr; <kbd>Formatting</kbd> and set <kbd>Formatter</kbd> to <kbd>External formatter</kbd>.

Does it solve your problem?

**Comment by boardend** [comment]
I have installed `pg_format` via the Ubuntu "pgformatter" package and it is working fine in Bash. But when I try to add it as an external formatter in DBeaver, I get `Cannot run program "pg_format": error=2, No such file or directory`. Could this be due to some sort of Snap sandboxing? (I'm on latest/stable 22.2.1.202209251341 from snapcraft.io)

**Comment by Destrolaric** [comment]
@boardend Yes, I tried this on DBeaver installed on snap and via non-sandboxing environments. This appears to be the snap issue. Thank you for the information.

**Comment by lav-patel** [comment]
> Hello @lav-patel,
> 
> You can define an external formatter in Preferences ⇒ Editors ⇒ SQL Editor ⇒ Formatting and set Formatter to External formatter.
> 
> Does it solve your problem?

It did work, but It made dbeaver unstable, so I had to revert the changes. I'll capture the error I get, and I put it here.

**Comment by theresa-de-ocampo** [comment]
> Hello @lav-patel,
> 
> You can define an external formatter in Preferences ⇒ Editors ⇒ SQL Editor ⇒ Formatting and set Formatter to External formatter.
> 
> Does it solve your problem?

@ShadelessFox Any progress update for the requested feature? On right click, it would ideally show something like `Format SQL (pgFormatter)`.

I tried to define pgFormatter as my external formatter by setting the `Command line` value to `C:\Strawberry\perl\site\bin\pg_format.bat ${file}`, and checking the `Use temp file`. It successfully formatted the preview.

![image](https://user-images.githubusercontent.com/80999430/235306218-d7e8d5f7-052a-4869-8fd5-d62497897db1.png)

But when I go to an SQL file, and run the formatter (_Ctrl+Shift+F_), it doesn't do anything. In fact, the execution is stuck as indicated in my status bar at the bottom of the window. I had to manually cancel the operation.

![pgFormatter](https://user-images.githubusercontent.com/80999430/235306563-d9634ad9-bce0-4455-850d-a92238914523.gif)

**Comment by lav-patel** [comment]
Yes it doesOn Apr 29, 2023, at 8:56 AM, Theresa De Ocampo ***@***.***> wrote:﻿

Hello @lav-patel,
You can define an external formatter in Preferences ⇒ Editors ⇒ SQL Editor ⇒ Formatting and set Formatter to External formatter.
Does it solve your problem?

Any progress update for the requested feature? On right click, it would ideally show something like Format SQL (pgFormatter).
I tried to define pgFormatter as my external formatter by setting the Command line value to C:\Strawberry\perl\site\bin\pg_format.bat ${file}, and checking the Use temp file. It successfully formatted the preview.

But when I go to an SQL file, and run the formatter (Ctrl+Shift+F), it doesn't do anything. In fact, the execution is stuck as indicated in my status bar at the bottom of the window. I had to manually cancel the operation.

**Comment by Reyesnes** [comment]
Hello.

I am trying to install pg-formatter in DBEAVER.
Is there any tutorial on how to do it? I have tried to follow the steps in the README, but it does not work for me.

I appreciate if someone can send me the step by step.

I have Windows 11

**Comment by hugorn** [comment]
> Hello.
> 
> I am trying to install pg-formatter in DBEAVER. Is there any tutorial on how to do it? I have tried to follow the steps in the README, but it does not work for me.
> 
> I appreciate if someone can send me the step by step.
> 
> I have Windows 11

Me too

**Comment by atr0000** [comment]
Freaking internat man ... i just lost multiple hours trying to find some good formatter for dbeaver that is free and does what i need ... god damn man the only answers i find is "dbeaver supports external formatter" or "there are free formatters online" or "read the documentation" but no one helps with installing these things step by step ... like i'm supposed to know another part of the programming/it/whatever world. shame on you guys for not spending a couple of minutes helping us out! i stopped looking and hpoing for an answer and asked chatgpt and installed it successfully. Though it would be nice to have multiple built in formatters inside dbeaver. 

Rant over. 

Because this is the top result when looking for dbeaver and pgformatter(at least for me) i'll provide here how i did it.

**Step 1:** Download and Install Perl from here https://strawberryperl.com/
Just go for latest release and simplest way is to download the msi (it's a windows installer you just double click next next finish)

To check if everything is ok open Command Prompt and write: "perl -v" if you get something then SUCCESS!!! if not ... u need to google the problem and start over.

**Step 2:** chatgpt says to install some perl modules because pgformatter needs them (i didn't so i didn't install but you may need them so here it is)
Open Command Prompt as Administrator and type:
cpan Text::CSV
cpan JSON::PP
cpan CGI

**Step 3:** Download and "install" pgformatter. Many options to download you'll find here: https://www.postgresql.org/about/news/pgformatter-v53-released-2493/

I just used sourceforge and got a zip file i unzipped it to "C:\tools\pgFormatter". Attention! In this folder is the "pg_format" file. You can unzip wherever you want.

To check if all is good you need to open Command Prompt go to the folder you have pgformatter like this:
"cd C:\tools\pgFormatter"
and then
"perl pg_format --version"

If you get something and not an error SUCCESS !!!! if there is an error please check the path 


**Step 3.1: ** 
This is optional but i usually do this for ease of use. You can skip it if you want. Add "C:\tools\pgFormatter" to the system environment path so when opening command prompt you can just do this wherever you are "pg_format --version" or if you want actually format something then like this "pg_format -f input.sql -o output.sql" 
NOTE: If this doesn't work go to your folder "C:\tools\pgFormatter" create a new bat (create it as txt and then change extension) file paste this: 
"
@echo off
perl C:\tools\pgFormatter\pg_format %*
"

**Step 4** 
Go to DBeaver then Window -> Preferences -> Editors -> SQL Editor -> Formatting then select external formatter and paste in the command line this: "perl C:\tools\pgFormatter\pg_format -" -> Apply -> Apply & Close.

Nothing more

All the options to configure the formatter are in the readme file:
https://github.com/darold/pgFormatter

If you don't get this then for example in dbeaver where you wrote "perl C:\tools\pgFormatter\pg_format -" for example if you want to have indent set at 8 you modify the line like this:
"perl C:\tools\pgFormatter\pg_format -s 8 -"

Furthermore if you want to add a newline after each comma then you modify the line like this:
"perl C:\tools\pgFormatter\pg_format -s 8 -B -"

If anyone want to jump in with more deatils please do and if you want to copy paste it and help others struggling please feel free to give them these steps.

Hope it helps the next poor soul trying to do this.

Edit: If you do all of this and still nothing go to Window -> Show View -> Error Log and see if you have any error when you try to format your code and debug from there.

Oh forgot to mention this is for windows ... if you need help on something else ask chatgpt or extrapolate the steps for your need.

**Comment by ShadelessFox** [comment]
@atr0000 thanks for the feedback.

We don't provide instructions because there are just too many options - and the steps for each of those differ drastically.

We can write an article covering popular options, such as the mentioned pgformatter.

**Comment by Matvey16** [comment]
Fixed for snap installation
